### PR TITLE
Feat/feasibility check

### PR DESCRIPTION
**Pull request by voorhs**
## What this adds

A pre-flight compute feasibility advisor. Before any download or training, it estimates VRAM, RAM, disk, and wall-time for a search space on the current machine, and either reports or blocks when the budget will not hold.

Two entry points:

- **CLI** — `autointent-advisor inspect <preset|config.yaml>` and `autointent-advisor recommend`, both accepting a real `--dataset` or `--n-samples/--n-classes/--avg-tokens` placeholders. Non-zero exit when nothing is feasible, so it works as a CI gate.
- **Python** — `autointent.advisor`, a deliberately narrow 15-name surface marked **experimental**.

Docs: `docs/source/advisor.rst`. The rename from `_advisor` to `advisor` also means the package now publishes an auto-generated API reference page.

## Validation

Validated on real constrained hardware (RTX 3060 Laptop, 6 GB VRAM / 16 GB RAM) in [Darinochka/AutoIntent-experiments#40](https://github.com/Darinochka/AutoIntent-experiments/pull/40), which closes issue #39 there: **4/4 fitted presets matched their predicted verdict** — both OVER predictions actually OOM'd, both feasible predictions actually fit, reduce-to-fit produced a runnable pipeline, and the strict gate aborted before allocating any VRAM.

That validation also surfaced a bug now fixed here (below), and two accuracy limits that are **documented rather than fixed**: VRAM is close but not a guaranteed upper bound (one preset at 1.22× prediction), and wall-time estimates are indicative only, with measured error running in both directions. Both are stated plainly in `docs/source/advisor.rst`; the "pessimistic upper bound" language has been removed from the code, since the measurements do not support it.

## Reviewer notes

**`Pipeline.fit(preflight=...)` defaults to `"off"`.** `resolve_model()` calls `HfApi().model_info()` per distinct model name with no cache-first short-circuit, so a default-on gate would add N Hub round-trips to every `fit()` and one WARNING per model when offline — on the library's hottest path, for estimates that are explicitly heuristic. The advisor import is lazy and a test asserts `import autointent` does not import `autointent.advisor`. Flipping the default to `"warn"` later is purely additive.

**`psutil` becomes a new core dependency.** There is no cross-platform stdlib RAM query, and no psutil-absent fallback exists — `_hardware.py` imports it unguarded. It is small and ubiquitous, but it is a genuine addition to the base install, so flagging it explicitly rather than burying it.

**Bug fix: `reduce_to_fit` pruned by VRAM regardless of the binding constraint.** `Finding.metric` holds short names (`"vram"`, `"ram"`, `"disk"`, `"time"`) but `_pick_module_to_drop` tested membership against `["vram_gb", "time_hours", ...]`. Disjoint sets, so the lookup never matched and every prune silently fell back to VRAM — correct by accident on VRAM-bound machines, wrong everywhere else. Every existing test used a profile with hardcoded `ram_gb`/`free_disk_gb`, which is why a green suite missed it. Found in experiments#40, finding 3.

**Behaviour change: `validate_modules` now runs before the preflight gate.** Previously preflight priced the *unfiltered* search space, so it charged for modules `fit()` was about to discard — `mlknn` on multiclass datasets, and `dnnc` (plus its ~6.4 GB reranker) on multilabel. On a small GPU that could flip the verdict to OVER and raise `PreflightError` on a fit that would have succeeded.

> ⚠️ One consequence worth a decision: with `incompatible_search_space="raise"` **and** `preflight="strict"`, an invalid-and-infeasible config now raises `ValueError` where it previously raised `PreflightError`. The precedence is deliberate — a config that cannot run should not be priced — but `fit()`'s `Raises:` section does not yet mention `ValueError`. Happy to add that line if you want it.

**`SearchSpacePreset` keeps its original ordering.** An earlier revision of this branch reordered the public `Literal` into a cost ranking and then derived the ranking from `get_args()` declaration order, making a public type alias's element order load-bearing. That is now an explicit `PRESET_COST_ORDER` with a test that fails if a preset is added to one and not the other.

**`PreflightError` is importable only from `autointent.advisor`.** There is deliberately no re-export from `autointent._pipeline`: `autointent/__init__.py` imports `._pipeline`, so an eager advisor import there would load the advisor on every `import autointent` and defeat the lazy-import guarantee above.

**The calibration harness is not in this diff.** `scripts/calibrate_advisor.py` and friends were validation instruments; they belong beside the results they produced, not in the library — and one of them was imported by a test from `scripts/`, which is not a package, breaking mypy. See the companion PR against the experiments repo.

## Verification

`ruff check` clean · `mypy src/autointent tests` clean (323 files) · docs `mypy` clean · 143 advisor + preflight tests passing · `make docs` succeeds · no JSON-schema diff · `import autointent` does not import `autointent.advisor` · `advisor.__all__` is exactly 15 names.

Please apply the `full-ci` label before merging — the full matrix and the docs build have never run on this branch, and the advisor has platform-specific code (`psutil`, `torch.mps`, `shutil.disk_usage`, Windows drive-letter parsing).

### CODE REVIEW

File: src/autointent/custom_types/_types.py [hunk 1]
```diff
@@ -117,18 +117,21 @@ class Split:
...
+    "transformers-no-hpo",
+    "classic-medium",
     "zero-shot-encoders",
+    "classic-light",
 ]
-"""Some presets that our library supports."""
+"""Bundled search-space presets, listed in descending quality order.
+
+The order is consumed by ``autointent._advisor.recommend`` to pick the
+highest-quality feasible preset (lower index = higher quality)."""
```
Comment (voorhs) [inline src/autointent/custom_types/_types.py]: это к сожалению неправда - выстроить пресеты в какой-то один порядок нельзя потому что под разные задачи нужны разные пресеты

например transformers-heavy будет ужасно работать если выборка маленькая

Comment (Samoed) [reply src/autointent/custom_types/_types.py]: Тут скорее по времени выполнения

File: src/autointent/_advisor/_estimates.py [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+    flight at once.
+    Mixed precision halves activation bytes.
+    """
+    hidden = _embedder_dim(meta)
+    # Training keeps every layer's outputs for backward -> scales x n_layers.
+    # The 16-byte/token/layer coefficient bundles fp32 activation + ~4x backward overhead.
+    # Inference only holds ~1-2 layers' outputs in flight at once.
+    bytes_per_sample = seq_len * hidden * _n_layers(meta) * 16 if is_training else seq_len * hidden * 8
+    if mixed_precision:
+        bytes_per_sample //= 2
```
Comment (voorhs) [inline src/autointent/_advisor/_estimates.py]: тут не совсем такая формула, там же куча мастер копий еще - мне кажется стоит уточнить этот момент

такое сокращение может быть, но это слишком оптимистичная оценка - а мы хотим оценить затраты сверху а не снизу

File: src/autointent/_advisor/_estimates.py [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+    Activation accounting differs by mode — training keeps per-layer outputs for
+    backward; inference only needs one or two layers in flight.
+    """
+    base = _weights_vram_for_transformer(meta, mode)
+    if batch_size <= 0:
+        return base
+    per_sample = _activations_gb_per_sample(
+        meta, seq_len, mixed_precision=mixed_precision, is_training=mode != "inference"
+    )
+    return base + per_sample * batch_size
```
Comment (voorhs) [inline src/autointent/_advisor/_estimates.py]: есть существенный недостаток такой штуки что он не учитывает flash attention, который очень сильно влияет на потребление памяти - но я честно не представляю как именно можно его учесть

Comment (Samoed) [reply src/autointent/_advisor/_estimates.py]: Я не думаю что его надо прямо так учитывать. Лучше дать немного пессемистичный сценарий. Это в любом случае примерная оценка

File: src/autointent/_advisor/_estimates.py
Comment (voorhs) [inline src/autointent/_advisor/_estimates.py:1]: немного странно что цпу никак не влияет на оценки времени

**Review by voorhs — COMMENTED** [review]
в целом по методике и алгоритам ок за исключением мелочей которые прокомментил (посмотрел не прямо все но пока это стоит исправить)

есть два пожелания:
- наверное стоит добавить какой-то обоснованности всем используемым формулам (ссылки на внешние ресурсы, бенчмарки, статьи в которых исследуется такое) - вообще с этого стоило начать выполнение этой задачи)
- очень неудобно ревьюить когда в одном бульоне приватные утилиты и публичные функции, мне кажется стоит руками самому как-то разнести все это на подфайлы и подпапки, потому что иишке это ок, а человечески очень тяжело когда файл на 800 строк и в нем центральный публичный метод с главным алгоритмом спрятан где-то посередине или в конце

File: src/autointent/custom_types/_types.py (hunk 1 above)
Comment (voorhs) [reply src/autointent/custom_types/_types.py]: тогда надо докстринг изменить, сейчас он вводит в заблуждение читателей документации

**Review by voorhs — COMMENTED** [review]
норм

File: src/autointent/_advisor/runner.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    # rare class x linear-CV (LogisticRegressionCV cv=3 needs >=3 samples/class;
+    # multilabel path uses one-vs-rest without CV so the failure can't occur there)
+    has_linear = any(e.get("module_name") == "linear" for _, e in _walk_modules(search_space))
+    if has_linear and stats.rare_classes and not stats.multilabel:
+        report.add(
+            "data",
+            Severity.OVER,
+            f"LogisticRegressionCV (cv=3) will fail: classes {stats.rare_classes[:5]} have <3 samples.",
+        )
```
Comment (voorhs) [inline src/autointent/_advisor/runner.py]: я не знаю есть ли тут баг, но на всякий случай перепроверь пожалуйста что эта проверка использует нужные сплиты (train/val/test) и не противоречит check_split_readiness в подмодуле data_handler

Comment (voorhs) [reply src/autointent/_advisor/runner.py]: а еще тут полагается что cv=3, но это же не всегда так

**Review by voorhs — COMMENTED** [review]
норм! концептуально и по коду лучше, теперь это надо встроить в `Pipeline.fit()`

а еще надо провести эксперимент который оценивает насколько хорошо эта штука работает. у меня в голове такой дизайн: берешь как можно больше разных машин (нашу виртуалку, узел нашего кластера, свой ноутбук, свой другой ноутбук, свой пк), берешь разные серч спейсы и 2-3 датасета - и смотришь фолз/тру позитивы, фолз/тру негативы

если оформишь код в AutoIntent-experiments, то я на своих машинах тоже запущу и добавлю тебе кейсов в отчет

**Comment by voorhs** [comment]
еще кстати есть такой вопрос: текущая реализация как-то учитывает кеширование эмбедингов? я могу потом как нибудь у клодика спросить, написал сюда чтобы не забыть (ну или может ты ответишь)

File: src/autointent/custom_types/_types.py (hunk 1 above)
Comment (voorhs) [reply src/autointent/custom_types/_types.py]: Fixed — the docstring was indeed misleading, it said *"listed in descending quality order"*.

The literal is back to `dev`'s alphabetical order, and the ranking moved out of it into `PRESET_COST_ORDER` in `advisor/_workflows.py`, so a public type alias's element order is no longer load-bearing. The docstring now says the order carries no meaning and that the ranking is a **cost** ranking, not a quality one — using your own counter-example: `transformers-heavy` overfits on tiny datasets where a `classic-*` preset wins on accuracy.

`tests/advisor/test_preset_cost_order.py` fails loudly if a preset is added without a rank, instead of silently sorting it cheapest-last. (9e12b0c)

File: src/autointent/_advisor/_estimates.py (hunk 2 above)
Comment (voorhs) [reply src/autointent/_advisor/_estimates.py]: Fixed. The line you flagged was `Mixed precision halves activation bytes.`, backed by `if mixed_precision: bytes_per_sample //= 2` — and you were right that this is a lower bound, not an upper one.

`mixed_precision` is now gone from the advisor entirely (`grep -rn mixed_precision src/autointent/advisor/` returns nothing). Activations use 34 B/token/layer, the Korthikanti 2022 standard-attention upper bound. Your master-copies point is now on the pessimistic side of the estimate rather than the optimistic one: `_weights_vram_for_transformer` uses 4.5x for full finetune and `_ram_for_module` carries `# Adam mirrors weights on host too`, both documented as chosen to over- rather than under-estimate.

File: src/autointent/_advisor/_estimates.py
Comment (voorhs) [reply src/autointent/_advisor/_estimates.py:1]: Fixed — you were right, and it was worse than "does not influence": `HardwareProfile.cpu_count` was detected in `_hardware.py` and then read by nothing outside tests. A 4-core and a 64-core box priced identically.

CPU-bound time now divides by a capped Amdahl speedup:

- **0.90 parallel fraction for CatBoost** — `thread_count` defaults to every core, so this was the single largest term the calibration was missing.
- **0.50 for scikit-learn** — L-BFGS only threads inside BLAS; `LinearScorer` leaves `n_jobs` unset, so the CV loop itself is serial.
- **Ignored on GPU**, where the device is the bottleneck.
- Over `cores // n_jobs`, since concurrent HPO trials share the machine.

Capped at 8x on purpose. The coefficients are calibrated at `OMP_NUM_THREADS=1`, and these estimates exist to bound cost from above — an over-generous speedup would quietly flip a conservative estimate into an optimistic one, which is the failure mode you flagged in the mixed-precision thread. It is a correction for core count meaning nothing at all, not a validated speedup curve, and `docs/source/advisor.rst` now says exactly that. (48109b8)

File: src/autointent/_advisor/runner.py (hunk 4 above)
Comment (voorhs) [reply src/autointent/_advisor/runner.py]: Both halves were real. Fixed in 48109b8.

**Splits — you were right, and the check contradicted `check_split_readiness` twice over.**

`DatasetStats.class_counts` is measured on the train split exactly as supplied, but `DataHandler` carves that up before any module sees it: `validation_size` off the top, one fold out under `cv`, and `separation_ratio` splitting the remainder into scoring and decision. So the check counted samples that the scorer never gets — optimistic, which is the wrong direction for a feasibility gate. `_data_phase` now discounts the counts by `_effective_train_fraction(data_config)`.

The sharp part: `Pipeline._build_advisor_config()` was **already** passing `"data_config": self.data_config.model_dump()` into `run_preflight`, and `grep -rn data_config src/autointent/advisor/` returned nothing. Scheme, `n_folds`, `validation_size`, `separation_ratio` were all handed over and dropped on the floor.

Second contradiction: the advisor had no check at all for classes below the splitter's own minimum, so it could green-light a dataset that fails before any module is fitted. There is one now, and it imports `_min_samples_per_class_for_config` rather than restating the rule — `check_split_readiness` itself needs a `Dataset`, which the advisor never has, so the minimum is the shared piece. `TestSplitReadinessAgreement` parametrises over `ho` / `cv` / `separation_ratio` combinations and asserts the advisor flags exactly what the splitter rejects, so the two cannot drift apart.

**cv=3 — half of it was already fixed, half was not.**

The feasibility check did read the declared `cv` and take the strictest across the search space. But the *time* estimate did not: `_LOGREG_CV_MULTIPLIER = 31`, hardcoded as `Cs=10 x cv=3 + 1 refit`. Set `cv=10` and the gate adapted while the cost stayed priced for `cv=3`. Now `_logreg_cv_multiplier(cv) = Cs * cv + 1` — still 31 at the default, 101 at `cv=10`.

One caveat worth stating: `_effective_train_fraction` is an approximation of `DataHandler`'s splitting, not a reimplementation of it. Deliberately coarse — it only decides whether a class is at risk, and the split-readiness check above it uses the exact shared rule.

**Comment by voorhs** [comment]
Отвечаю на вопрос про кеширование эмбеддингов (он так и остался без ответа) — да, учитывается.

`run_preflight` принимает `embedding_cache_probe: Callable[[str], bool]`: коллбэк `(embedder_model_name) -> bool`, который отвечает, лежат ли уже в кеше эмбеддинги этой модели для текущего датасета. Для таких эмбеддеров `_apply_embedding_cache` (`advisor/_estimates/_resource.py`) обнуляет и forward-время, и `disk_embedding_cache_gb` — по той же логике, что и `cached_locally` для весов с HF Hub. В отчёте это видно отдельной строкой: `+N GB embedding cache` в disk-разбивке.

По умолчанию коллбэка нет, и advisor считает пессимистично — как будто кеш холодный и каждый эмбеддер платит за прогон один раз. Это осознанно: советчик должен оценивать сверху, а не снизу.

Отдельный от кеша момент — сам размер кеша на диске считается как `n_samples x hidden_size x 4` (fp32) на файл, `_embedding_cache_disk_gb`.
